### PR TITLE
Analysis tools fix for label consistency between fragments and particles

### PR DESCRIPTION
**Pull request by dkoh0207**
1. Labels from `_fit_predict_groups` and `_fit_predict_interactions` are consistent with `ParticleFragment.group_id` and `ParticleFragment.interaction_id.
2. Labels between `Particle.id` and `ParticleFragment.group_id` are consistent
3. Labels between 'Particle.interaction_id` and `ParticleFragment.interaction_id` are consistent

### CODE REVIEW

File: mlreco/models/grappa.py [hunk 1]
```diff
@@ -200,13 +205,34 @@ def __init__(self, cfg, name='grappa', batch_col=0, coords_col=(1, 4)):
...
             node_output_feats = cfg[name]['gnn_model'].get('node_output_feats', 64)
-            vertex_config = cfg[name].get('vertex_net', {})
-            self.vertex_net = MomentumNet(node_output_feats, num_output=5, num_hidden=vertex_config.get('num_hidden', 128))
+            vertex_config = cfg[name].get('vertex_net', {'name': 'momentum_net'})
+            self.use_vtx_input_features = vertex_config.get('use_vtx_input_features', False)
+            if vertex_config['name'] == 'momentum_net':
+                self.vertex_net = MomentumNet(node_output_feats, 
+                                              num_output=5, 
+                                              num_hidden=vertex_config.get('num_hidden', 64),
+                                              evidential=True) # Enforce positive outputs
```
Comment (Temigo) [inline mlreco/models/grappa.py]: Does this mean evidential is enabled by default? does this change the previous default behavior of vertex prediction?

File: mlreco/models/layers/common/gnn_full_chain.py [hunk 2]
```diff
@@ -435,8 +436,8 @@ def run_particle_gnns(self, result, input, frag_result):
                            'node_pred_type': 'node_pred_type',
                            'node_pred_p': 'node_pred_p',
                            'node_pred_vtx': 'node_pred_vtx',
-                           'input_node_features': 'input_node_features',
-                           'input_edge_features': 'input_edge_features'}
+                           'input_node_features': 'particle_node_features',
+                           'input_edge_features': 'particle_edge_features'}
```
Comment (Temigo) [inline mlreco/models/layers/common/gnn_full_chain.py:440]: Why this renaming? Can you remind me what module(s) consume `input_node_features` and `input_edge_features` ?

File: mlreco/models/layers/gnn/factories.py [hunk 3]
```diff
@@ -142,8 +142,6 @@ def node_encoder_dict():
 
     encoders = {
         "geo"       : geometric.ClustGeoNodeEncoder,
-        "geo_norm"  : geometric.NormedClustGeoNodeEncoder,
```
Comment (Temigo) [inline mlreco/models/layers/gnn/factories.py:145]: Is this just completely obsolete by now? Can you clarify what it was/who was using it/why we remove it?

**Review by Temigo — COMMENTED** [review]
Please review my comments and confirm/clarify before we merge!

File: mlreco/models/grappa.py (hunk 1 above)
Comment (dkoh0207) [reply mlreco/models/grappa.py]: So `evidential` here is a vestigial name for changing the final output logit layer to pass through a softplus activation, which was needed to force `MomentumNet`'s output to be strictly positive.

We can change this to `positive_outputs=True` or the like

File: mlreco/models/layers/common/gnn_full_chain.py (hunk 2 above)
Comment (dkoh0207) [reply mlreco/models/layers/common/gnn_full_chain.py:440]: I've placed `input_xxxx_features` before for debugging purposes. I think we don't use input node features anywher except for vertex-net (which is under-construction anyway)

The name change was required because I needed to separate geometric encoder's output for 1) shower fragments 2) track fragments 3) particles (superset of fragments).

File: mlreco/models/layers/gnn/factories.py (hunk 3 above)
Comment (dkoh0207) [reply mlreco/models/layers/gnn/factories.py:145]: I think we agreed to just retrain the network on a larger ~6k^3 volume to resolve the spatial size dependency issue. Since I was the only one playing with it I just removed it

File: mlreco/models/grappa.py (hunk 1 above)
Comment (Temigo) [reply mlreco/models/grappa.py]: Okay, `positive_outputs` is probably a better name. My question remains: if it is set to `True` by default, it means the final output logit layer passes through a softplus activation by default. You say this was not the case before, correct? Does it mean the current weights everyone is using was trained with `evidential=False` ? I think if that is the case, it might throw off the vertex net predictions with current weights, right?

File: analysis/classes/ui.py [hunk 4]
```diff
@@ -87,12 +93,12 @@ def __init__(self, data_blob, result, cfg, predictor_cfg={}, deghosting=False):
         self.num_images = len(data_blob['input_data'])
         self.index = self.data_blob['index']
 
-        concat_result = cfg['trainval']['concat_result']
-        check_concat = set(self.CONCAT_RESULT)
-        for i, key in enumerate(concat_result):
-            if not key in check_concat:
-                raise ValueError("Output key <{}> should be listed under "\
-                    "trainval.concat_result!".format(key))
+        # concat_result = cfg['trainval']['concat_result']
```
Comment (Temigo) [inline analysis/classes/ui.py]: I disagree with this change. This means we have a default value for `concat_result` in two different places effectively. The `process_config` function already sets `cfg['trainval']['concat_result']` properly (to the default list of strings if necessary). Why does the analysis tools need to repeat the work of `process_config` ?

File: mlreco/models/layers/common/gnn_full_chain.py (hunk 2 above)
Comment (Temigo) [reply mlreco/models/layers/common/gnn_full_chain.py:440]: Okay that sounds good, thanks for clarifying

File: mlreco/models/grappa.py (hunk 1 above)
Comment (dkoh0207) [reply mlreco/models/grappa.py]: I think it won't crash when loading weights (since softplus doesn't have trainable weights) but if you have a trained vertex net and want to evaluate, yes that could be a problem. However did we have a working vertex-net at any point in time? (I can't remember)

File: analysis/classes/ui.py (hunk 4 above)
Comment (dkoh0207) [reply analysis/classes/ui.py]: Doesn't this change accomplish what you propose? (i.e., removing checking `concat_result` in two different places?) I'm seeing the code block `if not key in ...` being removed?

File: mlreco/models/layers/gnn/factories.py (hunk 3 above)
Comment (Temigo) [reply mlreco/models/layers/gnn/factories.py:145]: hmm, ok, I will clarify this with @francois-drielsma. My understanding was that we would do both, but I may very well misremember.

File: mlreco/models/grappa.py (hunk 1 above)
Comment (Temigo) [reply mlreco/models/grappa.py]: Not worried about the vertex coordinate prediction since that never worked. But we use it for primary particle prediction, if I am not mistaken. (The first 2 scores of the logits returned by vertex net)

File: analysis/classes/ui.py (hunk 4 above)
Comment (Temigo) [reply analysis/classes/ui.py]: Ah yes, my bad. In that case we might as well remove the attribute `CONCAT_RESULT` to avoid confusion?

File: mlreco/models/layers/gnn/factories.py (hunk 3 above)
Comment (Temigo) [reply mlreco/models/layers/gnn/factories.py:145]: Francois confirmed your memory that we are doing the retraining only for now, but he added that we might look more into spatial-size-agnostic features in the future? I am fine with removing this if you have no plans to extend it, but it's also fine to leave it for future reference.
